frappe.exceptions.ValidationError: As there are reserved stock, you cannot disable <strong>Stock Reservation</strong>.

### Information about bug


Test Case ID | Issue Type Recent
-- | --
TC_SCK_337 | frappe.throw(  File "/home/runner/frappe-bench/apps/frappe/frappe/__init__.py", line 637, in throw    msgprint(  File "/home/runner/frappe-bench/apps/frappe/frappe/__init__.py", line 602, in msgprint    _raise_exception()  File "/home/runner/frappe-bench/apps/frappe/frappe/__init__.py", line 553, in _raise_exception    raise excfrappe.exceptions.ValidationError: As there are reserved stock, you cannot disable <strong>Stock Reservation</strong>.



### Module

stock

### Version

Assets: v0.0.1 (pre-prod)

ERPNext: v15.60.0 (pre-prod)

ERPNext CRM: v0.0.1 (pre-prod)

ERPNext India: v0.0.1 (pre-prod)

Frappe Framework: v15.67.0 (pre-prod)

India Compliance: v15.17.1 (pre-prod)

Payments: v0.0.1 (version-15)

Projects: v0.0.1 (pre-prod)

Sales Commission: v0.0.1 (pre-prod)


### Installation method

None

### Relevant log output / Stack trace / Full Error Message.

```shell
"Traceback (most recent call last):
  File ""/home/runner/frappe-bench/apps/frappe/frappe/tests/utils.py"", line 313, in change_settings
    yield  # yield control to calling function
  File ""/opt/hostedtoolcache/Python/3.10.18/x64/lib/python3.10/contextlib.py"", line 79, in inner
    return func(*args, **kwds)
  File ""/home/runner/frappe-bench/apps/erpnext/erpnext/stock/doctype/stock_settings/test_stock_settings.py"", line 64, in test_validate_stock_reservation_TC_SCK_337
    with self.assertRaises(frappe.ValidationError, msg=msg):
AssertionError: ValidationError not raised : As there are negative stock, you can not enable Stock Reservation.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ""/opt/hostedtoolcache/Python/3.10.18/x64/lib/python3.10/contextlib.py"", line 78, in inner
    with self._recreate_cm():
  File ""/opt/hostedtoolcache/Python/3.10.18/x64/lib/python3.10/contextlib.py"", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File ""/home/runner/frappe-bench/apps/frappe/frappe/tests/utils.py"", line 320, in change_settings
    settings.save(ignore_permissions=True)
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 381, in save
    return self._save(*args, **kwargs)
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 417, in _save
    self.run_before_save_methods()
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 1140, in run_before_save_methods
    self.run_method(""validate"")
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 1011, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 1371, in composer
    return composed(self, method, *args, **kwargs)
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 1353, in runner
    add_to_return_value(self, fn(self, *args, **kwargs))
  File ""/home/runner/frappe-bench/apps/frappe/frappe/model/document.py"", line 1008, in fn
    return method_object(*args, **kwargs)
  File ""/home/runner/frappe-bench/apps/erpnext/erpnext/stock/doctype/stock_settings/stock_settings.py"", line 105, in validate
    self.validate_stock_reservation()
  File ""/home/runner/frappe-bench/apps/erpnext/erpnext/stock/doctype/stock_settings/stock_settings.py"", line 210, in validate_stock_reservation
    frappe.throw(
  File ""/home/runner/frappe-bench/apps/frappe/frappe/__init__.py"", line 637, in throw
    msgprint(
  File ""/home/runner/frappe-bench/apps/frappe/frappe/__init__.py"", line 602, in msgprint
    _raise_exception()
  File ""/home/runner/frappe-bench/apps/frappe/frappe/__init__.py"", line 553, in _raise_exception
    raise exc
frappe.exceptions.ValidationError: As there are reserved stock, you cannot disable <strong>Stock Reservation</strong>."
```

--- a/erpnext/stock/doctype/stock_settings/test_stock_settings.py
+++ b/erpnext/stock/doctype/stock_settings/test_stock_settings.py
@@ -55,15 +55,27 @@ class TestStockSettings(FrappeTestCase):
 	# codecov
 	@change_settings("Stock Settings", {"allow_negative_stock": 1, "enable_stock_reservation": 0})
 	def test_validate_stock_reservation_TC_SCK_337(self):
+		from frappe.query_builder.functions import Round
+
 		frappe.flags.in_test = False  # Force validation to run
 
 		settings = frappe.get_doc("Stock Settings")
 		settings.enable_stock_reservation = 1  # Trying to enable it
 		settings.allow_negative_stock = 0
 		msg = "As there are negative stock, you can not enable Stock Reservation."
-		with self.assertRaises(frappe.ValidationError, msg=msg):
+		precision = frappe.db.get_single_value("System Settings", "float_precision") or 3
+		bin = frappe.qb.DocType("Bin")
+		bin_with_negative_stock = (
+			frappe.qb.from_(bin).select(bin.name).where(Round(bin.actual_qty, precision) < 0).limit(1)
+		).run()
+		if bin_with_negative_stock:
+			msg = "As there are negative stock, you can not enable Stock Reservation."
+			with self.assertRaises(frappe.ValidationError, msg=msg):
+				settings.save()
+		else:
+			# In case there's no negative stock, it should save successfully
 			settings.save()
-		self.assertEqual(settings.enable_stock_reservation, 1)
+			self.assertEqual(settings.enable_stock_reservation, 1)
 
 	# codecov
 	@change_settings("Stock Settings", {"allow_negative_stock": 1, "enable_stock_reservation": 1})

--- a/erpnext/stock/doctype/stock_settings/test_stock_settings.py
+++ b/erpnext/stock/doctype/stock_settings/test_stock_settings.py
@@ -76,6 +76,9 @@ class TestStockSettings(FrappeTestCase):
 			# In case there's no negative stock, it should save successfully
 			settings.save()
 			self.assertEqual(settings.enable_stock_reservation, 1)
+		# reset config
+		settings.enable_stock_reservation = 0
+		settings.save()
 
 	# codecov
 	@change_settings("Stock Settings", {"allow_negative_stock": 1, "enable_stock_reservation": 1})

